### ISSUE TITLE
Support ECMAScript Modules in Pulumi JavaScript/TypeScript projects

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#4130]
Today, Pulumi Node.js programs are expected to be standard Node.js modules which can be loaded via `require()`.  The standard modules exports are then used as the stack exports, unless the full exported object itself is a function in which case that function is called with no arguments and if it returns a promise that promise is awaited.

ECMAScript modules are becoming slightly more common in Node.js, and we should have a story for supporting them.  That is, it should be possible to have the entrypoint of a Pulumi program be an ESM.

See https://nodejs.org/api/esm.html.

**Comment by lukehoban** [comment]
Initial support for this was added in https://github.com/pulumi/pulumi/pull/7764 (thanks @lewisl9029 🙌 ).  With that support, JavaScript and pre-compiled TypeScript Pulumi programs can be authored as ECMAScript modules by setting `"type": "module"` in `package.json` (and in the case of TypeScript, compiling with `module` and `target` settings that ensure modules are maintained, such as `"target": "ESNext"` and `"module": "ESNext"`).

Some additional support, in particular for automatic TypeScript support via `ts-node`, is coming in https://github.com/pulumi/pulumi/pull/8655, but the core support is available now, so closing this issue out. 

Here's an example of a configuration that now works for using ESM with Pulumi.

```json
{
    "name": "esm-js",
    "type": "module",
    "dependencies": {
        "@pulumi/aws": "^4.33.0",
        "@pulumi/pulumi": "^3.21.0"
    }
}
```

```js
import * as aws from "@pulumi/aws";

const buckets = [];
for (const az of (await aws.getAvailabilityZones()).names) {
    buckets.push(new aws.s3.Bucket(`my-bucket-${az}`));
}

export const bucketNames = buckets.map(b => b.id);
```

Note that the above example also takes advantage of [top-level await] which is supported in ECMAScript modules.

**Comment by joshua-barnett** [comment]
> Initial support for this was added in #7764 (thanks @lewisl9029 🙌 ). With that support, JavaScript and pre-compiled TypeScript Pulumi programs can be authored as ECMAScript modules by setting `"type": "module"` in `package.json` (and in the case of TypeScript, compiling with `module` and `target` settings that ensure modules are maintained, such as `"target": "ESNext"` and `"module": "ESNext"`).
> 
> Some additional support, in particular for automatic TypeScript support via `ts-node`, is coming in #8655, but the core support is available now, so closing this issue out.
> 
> Here's an example of a configuration that now works for using ESM with Pulumi.
> 
> ```json
> {
>     "name": "esm-js",
>     "type": "module",
>     "dependencies": {
>         "@pulumi/aws": "^4.33.0",
>         "@pulumi/pulumi": "^3.21.0"
>     }
> }
> ```
> 
> ```js
> import * as aws from "@pulumi/aws";
> 
> const buckets = [];
> for (const az of (await aws.getAvailabilityZones()).names) {
>     buckets.push(new aws.s3.Bucket(`my-bucket-${az}`));
> }
> 
> export const bucketNames = buckets.map(b => b.id);
> ```
> 
> Note that the above example also takes advantage of [top-level await] which is supported in ECMAScript modules.

```
$ pulumi about
CLI
Version      3.20.0
Go Version   go1.17.5
Go Compiler  gc

Plugins
NAME          VERSION
digitalocean  4.10.0
gitlab        4.4.0
kubernetes    3.12.1
nodejs        unknown

Host
OS       alpine
Version  3.14.3
Arch     x86_64

This project is written in nodejs (/usr/local/bin/node v16.13.0)

Current Stack: dev

TYPE                              URN
pulumi:pulumi:Stack               urn:pulumi:dev::awesome-todo::pulumi:pulumi:Stack::awesome-todo-dev
pulumi:providers:digitalocean     urn:pulumi:dev::awesome-todo::pulumi:providers:digitalocean::default_4_10_0
digitalocean:index/domain:Domain  urn:pulumi:dev::awesome-todo::digitalocean:index/domain:Domain::digitalocean-domain
pulumi:providers:gitlab           urn:pulumi:dev::awesome-todo::pulumi:providers:gitlab::default_4_4_0
gitlab:index/group:Group          urn:pulumi:dev::awesome-todo::gitlab:index/group:Group::gitlab-group


Found no pending operations associated with dev

Backend
Name     d2749b229d1b
URL      file://~
User     root

NAME                  VERSION
@pulumi/digitalocean  4.10.0
@pulumi/gitlab        4.4.0
@pulumi/kubernetes    3.12.1
@pulumi/pulumi        3.0.0
```


```
$ pulumi up
Diagnostics:
  pulumi:pulumi:Stack (awesome-todo-dev):
    error: Running program '/srv/todo/infra' failed with an unhandled exception:
    Error [ERR_REQUIRE_ESM]: require() of ES Module /srv/todo/infra/src/main.js from /tmp/awesome-todo/infra/node_modules/@pulumi/pulumi/cmd/run/run.js not supported.
    main.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
    Instead rename main.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /srv/todo/infra/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).
    
         at Object.<anonymous> (/tmp/awesome-todo/infra/node_modules/@pulumi/pulumi/cmd/run/run.js:214:31)
        at Generator.next (<anonymous>)
        at /tmp/awesome-todo/infra/node_modules/@pulumi/pulumi/cmd/run/run.js:21:71
        at new Promise (<anonymous>)
        at __awaiter (/tmp/awesome-todo/infra/node_modules/@pulumi/pulumi/cmd/run/run.js:17:12)
        at Object.runProgram [as init] (/tmp/awesome-todo/infra/node_modules/@pulumi/pulumi/cmd/run/run.js:200:30)
        at Stack.<anonymous> (/tmp/awesome-todo/infra/node_modules/@pulumi/pulumi/runtime/stack.js:86:43)
        at Generator.next (<anonymous>)
        at fulfilled (/tmp/awesome-todo/infra/node_modules/@pulumi/pulumi/runtime/stack.js:18:58
```

Doesn't appear to work.

**Comment by lukehoban** [comment]
> @pulumi/pulumi        3.21.0

This was not shipped in 3.21.0.  It will be available in the next release.  Or you can use `"@pulumi/pulumi": "dev",` to use the latest dev build, and it should work.
